ocaml/forkexecd/lib/fecomms.ml

```diff
@@ -23,12 +22,16 @@ let open_unix_domain_sock_client path =
 let read_raw_rpc sock =
   let buffer = Bytes.make 12 '\000' in
   Unixext.really_read sock buffer 0 12 ;
-  let len = int_of_string (Bytes.unsafe_to_string buffer) in
-  let body = Unixext.really_read_string sock len in
-  ferpc_of_rpc (Jsonrpc.of_string body)
+  let header = Bytes.unsafe_to_string buffer in
+  match int_of_string_opt header with
+  | Some len ->
+      let body = Unixext.really_read_string sock len in
+      Ok (Fe.ferpc_of_rpc (Jsonrpc.of_string body))
+  | None ->
+      Error ("Header is not an integer: " ^ header)
```

Comment:
If the RPC message is not valid, how can we trust that anything further we read from the socket will be valid? (the sender has broken the protocol)

We cannot close the socket here (it might lead to EBADF or sending data to the wrong socket), but we could shut it down such that any further attempts to read or write on the socket will fail immediately, see https://man7.org/linux/man-pages/man2/shutdown.2.html, in OCaml that would be `Unix.shutdown sock Unix.SHUTDOWN_ALL`

Reply:
Updated the reader to shut down the socket, the writer side was not dealing well with closed pipes either, added code for recovery or ignore the error where appropriate